lib/internal/modules/esm/resolve.js

```diff
@@ -506,9 +408,49 @@ function packageExportsResolve(
   throwExportsNotFound(packageSubpath, packageJSONUrl, base);
 }
 
+function readIfFile(path) {
+  let fd;
+  try {
+    fd = openSync(path, 'r');
+  } catch {
+    return undefined;
+  }
+  try {
+    if (!fstatSync(fd).isFile()) return undefined;
+    return readFileSync(fd, 'utf8');
+  } finally {
+    closeSync(fd);
+  }
+}
+
+function catchInvalidPackage(error, path, base) {
+  if (StringPrototypeIncludes(error.message, 'Error parsing')) {
```

Comment:
We could do an `error instanceof SyntaxError` here since the JSON error is the only case of a syntax error that can happen on the code path.